NotificationHubs/src/com/windowsazure/messaging/BaseInstallation.java

```diff
@@ -0,0 +1,310 @@
+    private String expirationTime;
+    private List<String> tags;
+    private Map<String, InstallationTemplate> templates;
+    private Map<String, WnsSecondaryTile> secondaryTiles;
+
+    /**
+    * Creates a new installation.
+    */
+    public BaseInstallation() {}
+
+    /**
+     * Creates a new installation with the given installation ID.
+     *
+     * @param installationId The ID for the installation.
+     */
+    public BaseInstallation(String installationId) {
+        this(installationId, (String[]) null);
+    }
+
+    /**
+     * Creates an installation from the Installation ID and tags.
+     *
+     * @param installationId The ID for the installation.
+     * @param tags           The tags for the installation.
+     */
+    public BaseInstallation(String installationId, String... tags) {
+        this(installationId, null, tags);
+    }
+
+    /**
+     * Creates an installation with an installation ID, platform and push channel.
+     *
+     * @param installationId The ID for the installation.
+     * @param platform       The platform for the installation.
+     */
+    public BaseInstallation(String installationId, NotificationPlatform platform) {
+        this(installationId, platform, (String[]) null);
+    }
+
+    public BaseInstallation(String installationId, NotificationPlatform platform, String... tags) {
+        // Validate that this is not FCM
+        validateNotificationPlatform(platform);
+
+        this.installationId = installationId;
+        this.platform = platform;
+        if (tags != null) {
+            for (String tag : tags) {
+                this.addTag(tag);
+            }
+        }
+    }
+
+    /**
+     * Validates that the platform is not FCM. Currently, Notification Hubs supports
+     * FCM Legacy as NotificationPlatform.Gcm Full support for FCM is not currently
+     * supported. See https://aka.ms/AA9dpaz
+     *
+     * @param notificationPlatform The notification platform to verify that it is
+     *                             not FCM.
+     */
+    private static void validateNotificationPlatform(NotificationPlatform notificationPlatform) {
+        if (notificationPlatform == NotificationPlatform.Fcm) {
+            throw new RuntimeException(
+                "FCM is currently not supported, use NotificationPlatform.Gcm which uses FCM Legacy Mode. See https://aka.ms/AA9dpaz");
+        }
+    }
+
+    /**
+     * Gets the ID for the installation.
+     *
+     * @return The ID for the installation.
+     */
+    public String getInstallationId() {
+        return installationId;
+    }
+
+    /**
+     * Sets the ID for the installation.
+     *
+     * @param value The ID for the installation.
+     */
+    public void setInstallationId(String value) {
+        installationId = value;
+    }
+
+    /**
+     * Gets whether the push channel has expired
+     *
+     * @return Returns true if expired, false otherwise.
+     */
+    public boolean isPushChannelExpired() {
```

Comment:
@louismagdaleno and I were going back and forth about whether this belongs with WNS Installations in particular, or if it's general enough to belong to the base Installation class. In general - my gut tells me to keep it specific to the types in which the PNS supports the notion of channel expiry, and that it should not be on the base.

Do you have a strong opinion either way?

Reply:
This belongs on all installations as you can set the expiration on every installation, just as we do for Android and Apple devices as part of extending the Installation expiration.  This is a getter only as it is serialized this way.